### PR TITLE
feat: Select

### PR DESCRIPTION
**Pull request by TingYinHelen**
- [x] Select Migration

### CODE REVIEW

**Comment by Glowdable** [comment]
Great job!

File: packages/select/src/option-group.vue [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    disabled: {
+      type: Boolean,
+      default: false,
+    },
+  },
+
+  setup(props) {
+    const visible = ref(true)
+
+    provide('SelectGroup', {
```
Comment (HerringtonDarkholme) [inline packages/select/src/option-group.vue:35]: toRefs

File: packages/select/src/option.vue [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+import { useOption } from './useOption'
+
+
+export default defineComponent({
+  name: 'ElOption',
+  componentName: 'ElOption',
+
+  props: {
+    value: {
+      required: true,
```
Comment (HerringtonDarkholme) [inline packages/select/src/option.vue:34]: Please fix the linting

File: packages/select/src/option.vue [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      // const { selected, multiple } = select
+      // let selectedOptions = multiple ? selected : [selected]
+      // let index = select.cachedOptions.value.indexOf(instance)
+      // let selectedIndex = selectedOptions?.indexOf(instance)
+
+      // if option is not selected, remove it from cache
+      // if (index > -1 && selectedIndex < 0) {
+      //   select.cachedOptions.value.splice(index, 1)
+      // }
+      select.onOptionDestroy(select.options.value.map(item => item.value).indexOf(props.value))
```
Comment (HerringtonDarkholme) [inline packages/select/src/option.vue:78]: `findIndex` ?

File: packages/select/src/option.vue [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    }
+  },
+
+  created() {
+    this.select.options.value.push(this)
+    this.select.cachedOptions.value.push(this)
+  },
+
+  methods: {
+    selectOptionClick() {
```
Comment (HerringtonDarkholme) [inline packages/select/src/option.vue:100]: Maybe we can declare it as VCA?

File: packages/select/src/select.vue [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+    id: String,
+    modelValue: {
+      type: [Array, String],
+    },
+    autocomplete: {
+      type: String,
+      default: 'off',
+    },
+    /** @Deprecated in next major version */
+    autoComplete: {
```
Comment (HerringtonDarkholme) [inline packages/select/src/select.vue:214]: just remove

File: packages/select/src/select-dropdown.vue [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  setup() {
+    const select = inject('Select', {} as any)
+
+    // computed
+    const popperClass = computed(() => select.props.popperClass)
+    const isMultiple = computed(() => select.props.multiple)
+    const minWidth = computed(() => select.selectWrapper.value && select.selectWrapper.value.getBoundingClientRect().width + 'px')
+
+    onMounted(() => {
+      // TODO: updatePopper
```
Comment (HerringtonDarkholme) [inline packages/select/src/select-dropdown.vue:32]: should popper be updated?

File: packages/select/src/select.vue [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+      menuVisibleOnFocus,
+      isOnComposition,
+      isSilentBlur,
+      options,
+      cachedOptions,
+      optionsCount,
+    } = toRefs(states)
+
+    // provide
+    provide('Select', {
```
Comment (HerringtonDarkholme) [inline packages/select/src/select.vue:354]: Why not provide state?

Comment (HerringtonDarkholme) [inline packages/select/src/select.vue:354]: Please use symbols and interface to type provide/injection

File: packages/select/src/useOption.ts [hunk 8]
```diff
@@ -0,0 +1,112 @@
+import {
+  inject,
+  computed,
+  getCurrentInstance,
+  watch,
+} from 'vue'
+import { getValueByPath, escapeRegexpString } from '@element-plus/utils/util'
+
+export const useOption = (props, states, ctx) => {
```
Comment (HerringtonDarkholme) [inline packages/select/src/useOption.ts:9]: Please fix the lint

File: packages/select/src/useOption.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+  // inject
+  const select = inject('Select', {}) as any
+  const selectGroup = inject('SelectGroup', {}) as any
+
+  // computed
+  const isObject = computed(() => {
```
Comment (HerringtonDarkholme) [inline packages/select/src/useOption.ts:15]: Could props be non object?

**Review by HerringtonDarkholme — APPROVED** [review]

**Pull request closed without merging**